Src/FluentAssertions/Equivalency/Matching/MustMatchByNameRule.cs

```diff
@@ -32,8 +32,13 @@ public IMember Match(IMember expectedMember, object subject, INode parent, IEqui
 
             if (subjectMember is null)
             {
-                Execute.Assertion.FailWith(
-                    $"Expectation has {expectedMember.Description} that the other object does not have.");
+                bool ignoreMemberBecauseItIsNotBrowsable = config.ExcludeNonBrowsable && !expectedMember.IsBrowsable;
+
+                if (!ignoreMemberBecauseItIsNotBrowsable)
```

Comment:
`IMemberMatchingRule`s are currently implemented as additive functions, i.e. if a member on `subject` matching `expectedMember` cannot be found, it will try the next rule.

See `StructuralEqualityEquivalencyStep.FindMatchFor`
```cs
private static IMember FindMatchFor(IMember selectedMember, INode currentNode, object subject,
    IEquivalencyAssertionOptions config)
{
    IEnumerable<IMember> query =
        from rule in config.MatchingRules
        let match = rule.Match(selectedMember, subject, currentNode, config)
        where match is not null
        select match;

    return query.FirstOrDefault();
}
```

What we want (it seems) here is to discard the `match` if either `selectedMember` or `match` is non-browsable.

So the check can not only be inside `MustMatchBeNameRule` as another `IMemberMatchingRule`, e.g. `TryMatchByNameRule` or even a user-provided `IMemberMatchingRule` could find a `match`.

Reply:
It seems that this breaks the nice abstraction & isolation of the design of the match rules. Now, before including a member from the subject, it needs to check for a matching rule in the expectation and vice versa. This is feeling like more code, more convoluted code, and less well-structured code. Are you sure it's preferable??

Reply:
Currently I don't have any idea how to solve the problem at all.
I just point out what does _not_ work.
Don't take it as critique, I'm just pressure testing the code to avoid releasing a half-baked feature.

Reply:
I would not expect the `ExcludingNonBrowsableMembers` to have any affect on the subject. If the expectation contains a property which is non-browsable, and you use the `ExcludingNonBrowsableMembers` option, it should just ignore that property and don't bother with it. Implementing that as a `IMemberSelectionRule` won't work either, since any of them can readd that property. I think you can solve this completely in the `StructuralEqualityEquivalencyStep` by first removing it from the final collection that `GetMembersFromExpectation` produces and then removing the subject property in `AssertMemberEquality` as you do now.

Reply:
Is that a preferable way to implement the feature? It seems neither offers perfect isolation, but the current implementation excludes both subject and expectation fields in the same `if` condition, and has one extra `if` to avoid complaining about missing members if they're excluded. If I'm understanding correctly, what you propose would use one mechanism for the expectation by post-processing the results of the selection rules, and then another, different mechanism for the subject. It should have the same effect but it feels more convoluted to me.

Reply:
In my humble opinion, it puts the logic where it belongs since these are two separate steps in the algorithm
1. Determine the members of the expectation to include
2. Find the matching member on the subject

Reply:
Hmm, okay. :-) I'll give it a go, see if I can't make it look like you're imagining.

Reply:
Okay that was a smaller change than I was expecting :-)

Reply:
Even better, I moved the subject check from `AssertMemberEquality` to `FindMatchFor` and it's now very symmetrical :-)

Reply:
But then you don't need these changes in `MustMatchByNameRule`, right?

Reply:
Indeed!